File: xplatform/lib/alignment/alignment_types.dart
import 'dart:convert';

/// One paragraph or chapter of source text addressable by stable id.
/// `id` matches an EPUB spine itemref so anchors round-trip across reimports.
class TextSegment {
  final String id;
  final String? title;
  final String text;

  const TextSegment({required this.id, this.title, required this.text});
}

class AlignmentInput {
  final List<TextSegment> segments;
  const AlignmentInput(this.segments);
}

class WordAnchor {
  final String segmentId;
  final int wordIndex;
  final double startSeconds;
  final double endSeconds;
  final int audioIndex;
  final double confidence;

  const WordAnchor({
    required this.segmentId,
    required this.wordIndex,
    required this.startSeconds,
    required this.endSeconds,
    this.audioIndex = -1,
    required this.confidence,
  });

  factory WordAnchor.fromJson(Map<String, dynamic> j) => WordAnchor(
        segmentId: j['segmentId'] as String,
        wordIndex: j['wordIndex'] as int,
        startSeconds: (j['startSeconds'] as num).toDouble(),
        endSeconds: (j['endSeconds'] as num).toDouble(),
        audioIndex: (j['audioIndex'] as int?) ?? -1,
        confidence: (j['confidence'] as num).toDouble(),
      );

  Map<String, dynamic> toJson() => {
        'segmentId': segmentId,
        'wordIndex': wordIndex,
        'startSeconds': startSeconds,
        'endSeconds': endSeconds,
        'audioIndex': audioIndex,
        'confidence': confidence,
      };
}

class SentenceAnchor {
  final String segmentId;
  final int sentenceIndex;
  final double startSeconds;
  final double endSeconds;

  const SentenceAnchor({
    required this.segmentId,
    required this.sentenceIndex,
    required this.startSeconds,
    required this.endSeconds,
  });

  factory SentenceAnchor.fromJson(Map<String, dynamic> j) => SentenceAnchor(
        segmentId: j['segmentId'] as String,
        sentenceIndex: j['sentenceIndex'] as int,
        startSeconds: (j['startSeconds'] as num).toDouble(),
        endSeconds: (j['endSeconds'] as num).toDouble(),
      );

  Map<String, dynamic> toJson() => {
        'segmentId': segmentId,
        'sentenceIndex': sentenceIndex,
        'startSeconds': startSeconds,
        'endSeconds': endSeconds,
      };
}

class AlignmentMap {
  final List<WordAnchor> words;
  final List<SentenceAnchor> sentences;
  /// Per-audio-word start timestamps from Whisper, in order. Index matches
  /// `WordAnchor.audioIndex` so playback can project audio time onto book
  /// position at narrator pace.
  final List<double> audioWordStarts;
  final DateTime createdAt;
  final String modelIdentifier;

  const AlignmentMap({
    required this.words,
    required this.sentences,
    this.audioWordStarts = const [],
    required this.createdAt,
    required this.modelIdentifier,
  });

  factory AlignmentMap.fromJson(Map<String, dynamic> j) => AlignmentMap(
        words: (j['words'] as List)
            .map((e) => WordAnchor.fromJson(e as Map<String, dynamic>))
            .toList(growable: false),
        sentences: (j['sentences'] as List)
            .map((e) => SentenceAnchor.fromJson(e as Map<String, dynamic>))
            .toList(growable: false),
        audioWordStarts: (j['audioWordStarts'] as List?)
                ?.map((e) => (e as num).toDouble())
                .toList(growable: false) ??
            const [],
        createdAt: DateTime.parse(j['createdAt'] as String),
        modelIdentifier: j['modelIdentifier'] as String,
      );

  factory AlignmentMap.fromJsonString(String s) =>
      AlignmentMap.fromJson(jsonDecode(s) as Map<String, dynamic>);

  Map<String, dynamic> toJson() => {
        'words': words.map((w) => w.toJson()).toList(growable: false),
        'sentences': sentences.map((s) => s.toJson()).toList(growable: false),
        'audioWordStarts': audioWordStarts,
        'createdAt': _iso8601Z(createdAt),
        'modelIdentifier': modelIdentifier,
      };

  String toJsonString() => jsonEncode(toJson());
}

/// Swift's JSONEncoder.dateEncodingStrategy = .iso8601 emits e.g.
/// "2026-05-07T17:42:57Z" — UTC, second precision, trailing Z. Match that
/// so files written on either platform round-trip identically.
String _iso8601Z(DateTime dt) {
  final u = dt.toUtc();
  String two(int n) => n.toString().padLeft(2, '0');
  return '${u.year.toString().padLeft(4, '0')}-${two(u.month)}-${two(u.day)}'
      'T${two(u.hour)}:${two(u.minute)}:${two(u.second)}Z';
}

File: PalimpsestCore/Sources/PalimpsestCore/Alignment/AlignmentTypes.swift
import Foundation

/// Input to an aligner: the ebook's text broken into addressable segments
/// (typically chapters or paragraphs). The `id` is a stable identifier
/// — for EPUB-sourced text this is an EPUB CFI; for plaintext it can be
/// any deterministic string.
public struct AlignmentInput: Sendable {
    public let segments: [TextSegment]
    public init(segments: [TextSegment]) { self.segments = segments }
}

public struct TextSegment: Sendable, Hashable {
    public let id: String
    public let title: String?
    public let text: String
    public init(id: String, title: String? = nil, text: String) {
        self.id = id
        self.title = title
        self.text = text
    }
}

/// The output of alignment: timestamps anchored back to the source text.
/// Both word- and sentence-level maps are produced from one alignment pass
/// so the reader can toggle granularity without re-running Whisper.
public struct AlignmentMap: Codable, Sendable {
    public let words: [WordAnchor]
    public let sentences: [SentenceAnchor]
    /// Per-audio-word start timestamps from Whisper, in order. Each entry's
    /// index matches the `audioIndex` used by `WordAnchor` so the reader can
    /// project audio time onto book word position at the narrator's actual
    /// pace instead of uniform time-to-word linear interpolation.
    public let audioWordStarts: [Double]
    public let createdAt: Date
    public let modelIdentifier: String

    public init(
        words: [WordAnchor],
        sentences: [SentenceAnchor],
        audioWordStarts: [Double] = [],
        createdAt: Date,
        modelIdentifier: String
    ) {
        self.words = words
        self.sentences = sentences
        self.audioWordStarts = audioWordStarts
        self.createdAt = createdAt
        self.modelIdentifier = modelIdentifier
    }

    private enum CodingKeys: String, CodingKey {
        case words, sentences, audioWordStarts, createdAt, modelIdentifier
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        self.words = try c.decode([WordAnchor].self, forKey: .words)
        self.sentences = try c.decode([SentenceAnchor].self, forKey: .sentences)
        self.audioWordStarts = (try? c.decode([Double].self, forKey: .audioWordStarts)) ?? []
        self.createdAt = try c.decode(Date.self, forKey: .createdAt)
        self.modelIdentifier = try c.decode(String.self, forKey: .modelIdentifier)
    }
}

public struct WordAnchor: Codable, Sendable, Hashable {
    public let segmentId: String
    public let wordIndex: Int
    public let startSeconds: Double
    public let endSeconds: Double
    /// Index into `AlignmentMap.audioWordStarts` where this anchor was matched.
    /// Lets the reader project audio time onto book wordIndex at narrator pace.
    public let audioIndex: Int
    /// 0.0 – 1.0. Below ~0.5, the reader should fall back to sentence highlighting.
    public let confidence: Float

    public init(
        segmentId: String,
        wordIndex: Int,
        startSeconds: Double,
        endSeconds: Double,
        audioIndex: Int = -1,
        confidence: Float
    ) {
        self.segmentId = segmentId
        self.wordIndex = wordIndex
        self.startSeconds = startSeconds
        self.endSeconds = endSeconds
        self.audioIndex = audioIndex
        self.confidence = confidence
    }

    private enum CodingKeys: String, CodingKey {
        case segmentId, wordIndex, startSeconds, endSeconds, audioIndex, confidence
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        self.segmentId = try c.decode(String.self, forKey: .segmentId)
        self.wordIndex = try c.decode(Int.self, forKey: .wordIndex)
        self.startSeconds = try c.decode(Double.self, forKey: .startSeconds)
        self.endSeconds = try c.decode(Double.self, forKey: .endSeconds)
        self.audioIndex = (try? c.decode(Int.self, forKey: .audioIndex)) ?? -1
        self.confidence = try c.decode(Float.self, forKey: .confidence)
    }
}

public struct SentenceAnchor: Codable, Sendable, Hashable {
    public let segmentId: String
    public let sentenceIndex: Int
    public let startSeconds: Double
    public let endSeconds: Double

    public init(segmentId: String, sentenceIndex: Int, startSeconds: Double, endSeconds: Double) {
        self.segmentId = segmentId
        self.sentenceIndex = sentenceIndex
        self.startSeconds = startSeconds
        self.endSeconds = endSeconds
    }
}
